content/blog/2024/2024-06-10-unit-test-for-architecture-using-archunit.md

```diff
@@ -0,0 +1,366 @@
+
+This article dives into ArchUnit, a powerful tool that lets us write automated unit tests for our software architecture.
+
+## What is ArchUnit
+
+ArchUnit is a free, simple and extensible library for checking the architecture of your Java code using any plain Java unit test framework. 
+
+ArchUnit has different layers: the Core layer, the Lang layer, and the Library layer.
+
+Much of the Core layer extends and closely resembles the Java reflection API. The Core layer offers a lot of information about the static structure of a Java program. For example, to import all Java classes in the com.akinwalehabib.archunitdemo package, we can use the ClassFileImporter from the Core layer:
+
+```java
+JavaClasses classes = new ClassFileImporter()
+  .importPackages("com.akinwalehabib.archunitdemo");
+```
+
+ArchUnit also has the Lang API, which provides a powerful syntax to express rules abstractly. Most parts of the Lang API are fluent APIs that are very expressive. You use the Lang API to create test rules. You can think of these rules as assertions in any unit test framework.
+
+```java
+ArchRule rule = classes()
+  .that()
+  .resideInAPackage("..service..")
+  .should()
+  .onlyBeAccessed()
+  .byAnyPackage("..controller..", "..service..");
+```
+
+In this code above, the rule states that all classes in the `service` package should only be accessible by classes in controller and the service package.
+
+Once we have imported the desired classes we wish to test and created our architecture rule using the Lang API, we check the rule against our imported classes:
+
+```java
+ArchRule rule = rule.check(importedClasses);
+```
+
+The Library API offers predefined complex rules for typical architectural goals. For example, we can use the Library API to check that a project adheres to a software architecture design such as the Onion architecture.
+
+In summary, we use the Objects from the Core API to import Java bytecode into Java Objects, then use Objects from the Lang API to specify our architecture rules. The Library API contains more complex predefined rules.
+
+## What to check
+
+What should we check in architectural tests in the first place? The typical checks we can do with ArchUnit are class dependencies and package dependency checks to enforce dependency rules in a class or between classes in different packages. Package dependency check is commonly used in Spring Modulith applications to check that classes do not depend on a class in another bounded context.
+
+## Getting started with ArchUnit
+
+Let us create a Spring Boot application. Although the example in this article is a bit contrived, it can help us understand how to use ArchUnit in a typical web application with different layers.
```

Comment:
The part of this article about creating a new Spring Boot application takes up over 50% of the article. This article is about ArchUnit and not Spring Boot. 

I think we can remove the whole start.spring.io stuff and just show some very simple example classes that have all the features you need for the use cases you're showing. Instead of the Spring Boot stuff, feel free to add more than the two example ArchUnit use cases.

I also suggest to put the code examples directly into the use case sections. I.e. In the section "Test to check repositories are only accessed by services", start with showing two very simple service and repository classes, and then show the ArchUnit test.